File: swift/Sources/PQCStandards/MlDsa/DsaNtt.swift
// ML-DSA Number Theoretic Transform (8 layers, n=256, q=8380417)

public enum DsaNtt {
    // Precomputed zetas: powers of 1753 mod q in bit-reversed order
    public static let zetas: [Int32] = {
        let q: Int64 = 8380417
        let g: Int64 = 1753
        var powers = [Int32](repeating: 0, count: 256)
        powers[0] = 1
        for i in 1..<256 {
            powers[i] = Int32((Int64(powers[i-1]) * g) % q)
        }
        func bitrev8(_ x: Int) -> Int {
            var r = 0; var v = x
            for _ in 0..<8 { r = (r << 1) | (v & 1); v >>= 1 }
            return r
        }
        var result = [Int32](repeating: 0, count: 256)
        for i in 0..<256 {
            result[i] = powers[bitrev8(i)]
        }
        return result
    }()

    /// Forward NTT (8 layers)
    public static func ntt(_ f: inout [Int32]) {
        var k = 1
        var len = 128
        while len >= 1 {
            var start = 0
            while start < 256 {
                let z = zetas[k]
                k += 1
                for j in start..<(start + len) {
                    let t = DsaField.mul(z, f[j + len])
                    f[j + len] = DsaField.sub(f[j], t)
                    f[j] = DsaField.add(f[j], t)
                }
                start += 2 * len
            }
            len /= 2
        }
    }

    /// Inverse NTT (8 layers)
    public static func invNtt(_ f: inout [Int32]) {
        var k = 255
        var len = 1
        while len <= 128 {
            var start = 0
            while start < 256 {
                let z = zetas[k]
                k -= 1
                for j in start..<(start + len) {
                    let t = f[j]
                    f[j] = DsaField.add(t, f[j + len])
                    f[j + len] = DsaField.mul(z, DsaField.sub(f[j + len], t))
                }
                start += 2 * len
            }
            len *= 2
        }
        // Multiply by n^{-1} mod q
        // 256^{-1} mod 8380417 = 8347681
        let nInv: Int32 = 8347681
        for i in 0..<256 {
            f[i] = DsaField.mul(f[i], nInv)
        }
    }

    /// Pointwise multiply (NTT domain)
    public static func pointwiseMul(_ a: [Int32], _ b: [Int32]) -> [Int32] {
        var r = [Int32](repeating: 0, count: 256)
        for i in 0..<256 {
            r[i] = DsaField.mul(a[i], b[i])
        }
        return r
    }

    /// Add two polynomials
    public static func polyAdd(_ a: [Int32], _ b: [Int32]) -> [Int32] {
        var r = [Int32](repeating: 0, count: 256)
        for i in 0..<256 {
            r[i] = DsaField.add(a[i], b[i])
        }
        return r
    }

    /// Subtract two polynomials
    public static func polySub(_ a: [Int32], _ b: [Int32]) -> [Int32] {
        var r = [Int32](repeating: 0, count: 256)
        for i in 0..<256 {
            r[i] = DsaField.sub(a[i], b[i])
        }
        return r
    }
}

File: php/src/MlDsa/DsaNtt.php
<?php

declare(strict_types=1);

namespace PQC\MlDsa;

/**
 * NTT for ML-DSA (q=8380417, zeta=1753).
 */
final class DsaNtt
{
    private static ?array $zetas = null;

    /**
     * Precompute zetas: zeta^{BitRev8(i)} mod q.
     */
    public static function getZetas(): array
    {
        if (self::$zetas !== null) {
            return self::$zetas;
        }

        // Precomputed zetas for ML-DSA
        $zeta = 1753;
        $q = DsaField::Q;

        // Compute powers of zeta in bit-reversed order
        $powers = [];
        $powers[0] = 1;
        for ($i = 1; $i < 256; $i++) {
            $powers[$i] = DsaField::mul($powers[$i - 1], $zeta);
        }

        self::$zetas = [];
        for ($i = 0; $i < 256; $i++) {
            $br = self::bitRev8($i);
            self::$zetas[$i] = $powers[$br];
        }

        return self::$zetas;
    }

    private static function bitRev8(int $n): int
    {
        $r = 0;
        for ($i = 0; $i < 8; $i++) {
            $r = ($r << 1) | ($n & 1);
            $n >>= 1;
        }
        return $r;
    }

    /**
     * Forward NTT.
     */
    public static function ntt(array $f): array
    {
        $zetas = self::getZetas();
        $k = 1;
        $len = 128;

        while ($len >= 1) {
            for ($start = 0; $start < 256; $start += 2 * $len) {
                $z = $zetas[$k++];
                for ($j = $start; $j < $start + $len; $j++) {
                    $t = DsaField::mul($z, $f[$j + $len]);
                    $f[$j + $len] = DsaField::sub($f[$j], $t);
                    $f[$j] = DsaField::add($f[$j], $t);
                }
            }
            $len >>= 1;
        }

        return $f;
    }

    /**
     * Inverse NTT.
     */
    public static function invNtt(array $f): array
    {
        $zetas = self::getZetas();
        $k = 256;
        $len = 1;

        while ($len <= 128) {
            for ($start = 0; $start < 256; $start += 2 * $len) {
                $k--;
                $z = $zetas[$k];
                for ($j = $start; $j < $start + $len; $j++) {
                    $t = $f[$j];
                    $f[$j] = DsaField::add($t, $f[$j + $len]);
                    $f[$j + $len] = DsaField::mul($z, DsaField::sub($f[$j + $len], $t));
                }
            }
            $len <<= 1;
        }

        $nInv = DsaField::inv(256);
        for ($i = 0; $i < 256; $i++) {
            $f[$i] = DsaField::mul($f[$i], $nInv);
        }

        return $f;
    }

    /**
     * Pointwise multiplication in NTT domain.
     */
    public static function mulNtt(array $a, array $b): array
    {
        $c = [];
        for ($i = 0; $i < 256; $i++) {
            $c[$i] = DsaField::mul($a[$i], $b[$i]);
        }
        return $c;
    }

    /**
     * Polynomial addition.
     */
    public static function polyAdd(array $a, array $b): array
    {
        $c = [];
        for ($i = 0; $i < 256; $i++) {
            $c[$i] = DsaField::add($a[$i], $b[$i]);
        }
        return $c;
    }

    /**
     * Polynomial subtraction.
     */
    public static function polySub(array $a, array $b): array
    {
        $c = [];
        for ($i = 0; $i < 256; $i++) {
            $c[$i] = DsaField::sub($a[$i], $b[$i]);
        }
        return $c;
    }

    /**
     * Matrix-vector product: A * v (both in NTT domain).
     */
    public static function matVecMul(array $mat, array $vec, int $k, int $l): array
    {
        $result = [];
        for ($i = 0; $i < $k; $i++) {
            $result[$i] = array_fill(0, 256, 0);
            for ($j = 0; $j < $l; $j++) {
                $prod = self::mulNtt($mat[$i][$j], $vec[$j]);
                $result[$i] = self::polyAdd($result[$i], $prod);
            }
        }
        return $result;
    }

    /**
     * Vector dot product in NTT domain.
     */
    public static function vecDot(array $a, array $b, int $len): array
    {
        $result = array_fill(0, 256, 0);
        for ($i = 0; $i < $len; $i++) {
            $prod = self::mulNtt($a[$i], $b[$i]);
            $result = self::polyAdd($result, $prod);
        }
        return $result;
    }
}
